### DISCUSSION TITLE
Convert JS to typescript

### DISCUSSION THREAD

**Discussion by mnahkies** [invertase/react-native-firebase#7568] [Ideas]
Currently the typescript definitions are written by hand, and feel a bit awkward to consume (IMO). There is some limited use of typescript for writing the `expo` config plugin already.

Is there any appetite to put the structure in place to start writing the JS side of the modules in typescript?

**Comment by mikehardy** [comment]
Definitely appetite - I'd personally love to see it in typescript as writing in javascript always makes me feel like I'm on a tightrope without a safety net

I haven't personally explored what is required but *I imagine* (big warning this is all untested theory...) it is a matter of setting up the typescript config file correctly, implementing something along the lines of react-native-builder-bob to do the build, re-pointing various things to go to the built lib vs src, and...if things are set up well for incremental conversion, that *should* be sufficient to start? As in it should be running even before the first file is converted?

At that point our conversion processes usually go from easiest to hardest, where difficulty is judged by code size + complexity and test complexity. Easiest is ML (because it almost doesn't exist! it's effectively empty) followed most likely by in-app-messaging as the smallest modules. Then probably remote-config. 

If those are working well, then it's just a matter of grinding down the work on rest of the modules
